### ISSUE TITLE
Unclosed character class / Java Exception

### ISSUE THREAD

**Issue by WIRUT** [theY4Kman/pycharm-pytest-imp#6]
I'm getting an error where it looks like the regex rebuild/conversion is incorrectly parsing. My Pytest.ini is small, containing only:
```
[pytest]
python_classes = C[0-9][0-9][0-9[0-9][0-9]
addopts = --cache-clear
python_functions = test2* *test2
usefixtures = config
```

This is what shows up in my event log
```
During querying provider Run line marker (class com.intellij.execution.lineMarker.RunLineMarkerProvider)

java.util.regex.PatternSyntaxException: Unclosed character class near index 59
C[0[A-Z0-9]9][0[A-Z0-9]9][0[A-Z0-9]9[0[A-Z0-9]9][0[A-Z0-9]9]
                                                           ^
	at java.base/java.util.regex.Pattern.error(Pattern.java:2027)
	at java.base/java.util.regex.Pattern.clazz(Pattern.java:2695)
	at java.base/java.util.regex.Pattern.sequence(Pattern.java:2138)
	at java.base/java.util.regex.Pattern.expr(Pattern.java:2068)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1782)
	at java.base/java.util.regex.Pattern.<init>(Pattern.java:1428)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1068)
	at kotlin.text.Regex.<init>(Regex.kt:89)
	at com.y4kstudios.pytestimp.PyTestImpServiceKt.convertWildcardPatternsToRegex(PyTestImpService.kt:109)
	at com.y4kstudios.pytestimp.PyTestConfig$pythonClasses$2.invoke(PyTestImpService.kt:119)
	at com.y4kstudios.pytestimp.PyTestConfig$pythonClasses$2.invoke(PyTestImpService.kt:115)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at com.y4kstudios.pytestimp.PyTestConfig.getPythonClasses(PyTestImpService.kt)
	at com.y4kstudios.pytestimp.runconfigs.PyTestLineMarkerContributorKt.isPytestIniConfiguredTestElement(PyTestLineMarkerContributor.kt:43)
	at com.y4kstudios.pytestimp.runconfigs.PyTestLineMarkerContributor.getInfo(PyTestLineMarkerContributor.kt:28)
	at com.intellij.execution.lineMarker.RunLineMarkerProvider.getLineMarkerInfo(RunLineMarkerProvider.java:50)
	at com.intellij.codeInsight.daemon.impl.LineMarkersPass.queryProviders(LineMarkersPass.java:158)
	at com.intellij.codeInsight.daemon.impl.LineMarkersPass.lambda$doCollectInformation$3(LineMarkersPass.java:83)
	at com.intellij.codeInsight.daemon.impl.Divider.divideInsideAndOutsideInOneRoot(Divider.java:81)
	at com.intellij.codeInsight.daemon.impl.LineMarkersPass.doCollectInformation(LineMarkersPass.java:78)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:56)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$1(PassExecutorService.java:400)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:393)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:392)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:368)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:183)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:366)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:188)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by theY4Kman** [comment]
Oops, yeah, I'm pretty sure this is due to my wonky support for [pytest-camel-collect](https://pypi.org/project/pytest-camel-collect/), whereby dashes are literally replaced with `[A-Z0-9]`, regardless of their location in the string. Fixing it ought to be pretty simple, by substituting the simple replace with a regular expression.

In the long term, I'd like to probe the Python environment for existence of the pytest-camel-collect package to enable/disable that support — but that's irrelevant for this issue.

**Comment by theY4Kman** [comment]
Okay, I have a fix for incorrect substitution of dashes, but it appears the `python_classes` you've shown above is missing the closing `]` bracket, so even when properly substituted, it still breaks. I'll come up with a better solution for that, and hopefully will publish this fix soon. It's already way past bedtime :P

**Comment by WIRUT** [comment]
Ah whoops! Copy error. But thanks for looking into it!

**Comment by theY4Kman** [comment]
Okie dokes, I ended up writing a little fnmatch parser-transformer to implement the CamelCase word boundary dashes correctly. The upshot of which is that the plugin now properly parses `?` single-character matches and `[!abc]` negative character classes in the `python_classes` and `python_functions` patterns. Yay! :D

Oh, and I also made it handle unclosed character classes gracefully by treating such patterns as "match nothing". Ideally, the plugin would add annotations (i.e. red squiggly lines and such) to the `pyproject.toml` or `pytest.ini` source of the faulty patterns, but that'll have to come in the future.

I just published version 0.4.2 to the JetBrains plugin repo, and it should be available within a day or two. I've also uploaded the plugin .zip to the [0.4.2 GitHub release](https://github.com/theY4Kman/pycharm-pytest-imp/releases/tag/0.4.2), should you wish to test it out sooner.

Thanks for reporting this! Beyond the dash fix, it gave me the opportunity to greatly improve the fnmatch implementation, for the benefit of all :)

I'm gonna close this for now, but holler at me if my fix falls short :P
